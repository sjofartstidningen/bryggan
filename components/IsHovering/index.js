import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class IsHovering extends Component {
  static propTypes = {
    children: PropTypes.func.isRequired,
    onMouseEnter: PropTypes.func,
    onMouseLeave: PropTypes.func,
  };

  static defaultProps = {
    onMouseEnter: null,
    onMouseLeave: null,
  };

  state = {
    hovering: false,
  };

  handleMouseEnter = () => {
    this.setState(() => ({ hovering: true }));
    if (typeof this.props.onMouseEnter === 'function') {
      this.props.onMouseEnter();
    }
  };

  handleMouseLeave = () => {
    this.setState(() => ({ hovering: false }));
    if (typeof this.props.onMouseLeave === 'function') {
      this.props.onMouseLeave();
    }
  };

  render() {
    const { hovering } = this.state;
    const { children, ...props } = this.props;
    return (
      <div
        {...props}
        onMouseEnter={this.handleMouseEnter}
        onMouseLeave={this.handleMouseLeave}
      >
        {children({ hovering })}
      </div>
    );
  }
}
