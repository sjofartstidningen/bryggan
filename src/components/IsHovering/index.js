import React, { Component } from 'react';
import PropTypes from 'prop-types';

class IsHovering extends Component {
  static propTypes = {
    el: PropTypes.any, // eslint-disable-line react/forbid-prop-types
    render: PropTypes.func.isRequired,
    className: PropTypes.string,
  };

  static defaultProps = {
    el: 'div',
    className: '',
  };

  state = {
    isHovering: false,
  };

  handleMouseEnter = () => {
    this.setState(() => ({ isHovering: true }));
  };

  handleMouseLeave = () => {
    this.setState(() => ({ isHovering: false }));
  };

  render() {
    const { el: El, render, className, ...props } = this.props;
    const { isHovering } = this.state;

    return (
      <El
        {...props}
        className={className}
        onMouseEnter={this.handleMouseEnter}
        onMouseLeave={this.handleMouseLeave}
      >
        {render({ isHovering })}
      </El>
    );
  }
}

export default IsHovering;
