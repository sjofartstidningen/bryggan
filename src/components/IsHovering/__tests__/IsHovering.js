import React from 'react';
import { mount } from 'enzyme';
import IsHovering from '../';

function setup() {
  const renderSpy = jest.fn(({ isHovering }) => (
    <div className="inner" style={{ color: isHovering ? 'red' : 'black' }} />
  ));

  const Component = props => (
    <IsHovering className="outer" {...props} render={renderSpy} />
  );

  return { Component, renderSpy };
}

describe('Component.IsHovering', () => {
  test('should react to onMouseEnter/Leave events', () => {
    const { Component, renderSpy } = setup();
    const wrapper = mount(<Component />);

    expect(renderSpy).toHaveBeenCalledTimes(1);

    const outer = wrapper.find('.outer').first();
    outer.simulate('mouseenter', { mouseenter: true });
    expect(renderSpy).toHaveBeenCalledTimes(2);

    outer.simulate('mouseleave', { mouseleave: true });
    expect(renderSpy).toHaveBeenCalledTimes(3);
  });

  test('should update render prop when mouse hovers', () => {
    const { Component } = setup();
    const wrapper = mount(<Component />);
    const outer = wrapper.find('.outer').first();

    expect(wrapper).toMatchSnapshot();
    outer.simulate('mouseenter', { mouseenter: true });
    expect(wrapper).toMatchSnapshot();
  });
});
