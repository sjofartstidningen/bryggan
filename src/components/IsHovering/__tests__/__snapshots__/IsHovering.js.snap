// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component.IsHovering should update render prop when mouse hovers 1`] = `
<Component>
  <IsHovering
    className="outer"
    el="div"
    render={[Function]}
  >
    <div
      className="outer"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="inner"
        style={
          Object {
            "color": "black",
          }
        }
      />
    </div>
  </IsHovering>
</Component>
`;

exports[`Component.IsHovering should update render prop when mouse hovers 2`] = `
<Component>
  <IsHovering
    className="outer"
    el="div"
    render={[Function]}
  >
    <div
      className="outer"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="inner"
        style={
          Object {
            "color": "red",
          }
        }
      />
    </div>
  </IsHovering>
</Component>
`;
